cpp/test/sg/kpca_test.cu

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2018-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include <raft/cudart_utils.h>
+#include <raft/linalg/cublas_wrappers.h>
+#include <test_utils.h>
+#include <cuml/decomposition/params.hpp>
+#include <kpca/kpca.cuh>
+#include <raft/cuda_utils.cuh>
+#include <raft/random/rng.cuh>
+#include <vector>
+
+namespace ML {
+
+using namespace MLCommon;
+
+template <typename T>
+struct KPcaInputs {
+  T tolerance;
+  int len;
+  int n_row;
+  int n_col;
+  int len2;
+  int n_row2;
+  int n_col2;
+  unsigned long long int seed;
+  int algo;
+};
+
+template <typename T>
+::std::ostream& operator<<(::std::ostream& os, const KPcaInputs<T>& dims) {
+  return os;
+}
+
+template <typename T>
+class KPcaTest : public ::testing::TestWithParam<KPcaInputs<T>> {
+ protected:
+  void linearTest() {
+    params = ::testing::TestWithParam<KPcaInputs<T>>::GetParam();
+    raft::random::Rng r(params.seed, raft::random::GenTaps);
+    int len = params.len;
+
+    raft::allocate(data, len);
+    raft::allocate(data_back, len);
+    raft::allocate(trans_data, len); // transformed data
+
+    std::vector<T> data_h = {1.0, 2.0, 5.0, 4.0, 2.0, 1.0};
```

Comment:
Make host vectors part of input struct for all

Reply:
Yeah.  I kludged the tests to make sure I could run linear, polynomial, and RBF tests.  I will streamline it.